hie-plugin-api/Haskell/Ide/Engine/Context.hs

```diff
@@ -13,27 +15,86 @@ import Haskell.Ide.Engine.PluginUtils
 -- smarter code completion
 data Context = TypeContext
              | ValueContext
+             | ModuleContext String
+             | ImportContext String
+             | ImportListContext String
+             | ImportHidingContext String
+             | ExportContext
+             | InstanceContext
+             | ClassContext
+             | DerivingContext
   deriving (Show, Eq)
 
 -- | Generates a map of where the context is a type and where the context is a value
 -- i.e. where are the value decls and the type decls
 getContext :: Position -> ParsedModule -> Maybe Context
-getContext pos pm = everything join (Nothing `mkQ` go `extQ` goInline) decl
+getContext pos pm
+  | Just (L (RealSrcSpan r) modName) <- moduleHeader
+  , pos `isInsideRange` r
+  = Just (ModuleContext (moduleNameString modName))
+
+  | Just (L (RealSrcSpan r) _) <- exportList
+  , pos `isInsideRange` r
+  = Just ExportContext
+
+  | Just ctx <- everything (<|>) (Nothing `mkQ` go `extQ` goInline) decl
+  = Just ctx
+
+  | Just ctx <- asum $ map importGo imports
+  = Just ctx
+
+  | otherwise
+  = Nothing
+
   where decl = hsmodDecls $ unLoc $ pm_parsed_source pm
+        moduleHeader = hsmodName $ unLoc $ pm_parsed_source pm
+        exportList = hsmodExports $ unLoc $ pm_parsed_source pm
+        imports = hsmodImports $ unLoc $ pm_parsed_source pm
+
         go :: LHsDecl GM.GhcPs -> Maybe Context
-        go (L (RealSrcSpan r) (SigD {}))
+        go (L (RealSrcSpan r) SigD {})
           | pos `isInsideRange` r = Just TypeContext
           | otherwise = Nothing
-        go (L (GHC.RealSrcSpan r) (GHC.ValD {}))
+        go (L (GHC.RealSrcSpan r) GHC.ValD {})
           | pos `isInsideRange` r = Just ValueContext
           | otherwise = Nothing
+        go (L (GHC.RealSrcSpan r) GHC.InstD {})
+          | pos `isInsideRange` r = Just InstanceContext
+          | otherwise = Nothing
+        go (L (GHC.RealSrcSpan r) GHC.DerivD {})
+          | pos `isInsideRange` r = Just DerivingContext
+          | otherwise = Nothing
+        go (L (GHC.RealSrcSpan r) (GHC.TyClD _ GHC.ClassDecl {}))
```

Comment:
How should we handle GHC AST version differences?
A quick work around would be, to add another view pattern to `Compat`.
However, it seems like this would grow out of control quickly. Should we rather use something more version independent, such as ghc-lib or haskell-src-exts?